### PR TITLE
[Refactor] Add ExploreGroupTask in memo phase

### PR DESCRIPTION
**Pull request by Seaven**
## What type of PR is this：
- [ ] bug
- [ ] feature
- [ ] enhancement
- [x] refactor
- [ ] others

## Which issues of this PR fixes ：

Fixes #

## Problem Summary(Required) ：

Ok, let's use our imagination.

For a complex SQL (select * from A join B join C join D join E). In Memo, we found two logical plan:
```
# 1
        GroupExpression(AxBxC) x (DxE)
           /             \
      Group(AxBxC)      Group(DxE)
          |                |
        .....            ......

# 2
        GroupExpression(AxBxCxD) x (E)
           /             \
      Group(AxBxCxD)     Group(E)
          |                 |
        .....            ......
```

The optimizer process now:
1. We will compute all statistics of logical node before enter memo.
2. We optimize expression (AxBxC)x(DxE), enforce it,  and get cost-1, then save the best expression 
3. Then we update the plan upper bound by cost-1, and to enforce (AxBxCxD) x (E)
4. We use the statistics of (AxBxCxD) to compute the cost of Group(AxBxCxD), then check is greate upper bound cost-1, and prune the plan if cost greate upper bound.

Ok, question is the statistics of Group(AxBxCxD) never updated, because we have been derive the statistics of this group from the beginning. We only re-dervie statistics when produce new logical expression.
But in actual, origin GroupExpression (AxBxCxD) maybe is (A cross join B cross join C cross join D), and it's will produce new GroupExpression (A inner join D inner join B inner join C) when apply transform rule. Obviously the cost of the latter is lower than the former, but we can't get the latter, because Group(AxBxCxD) has been pruned.

the reason:
On paper, no matter how to change Child-Plan(ABCD), the statistics of Goup(ABCD)  is always the same because his output doesn't change.But the algorithm of statistics can't promise idempotency, such as: the output of (A InnerJoin B on A.a = B.b) is different with the output of (A CrossJoin B on A.a = B.b), but their outputs are same in actual, it's will cause some complex plans to be discarded because the original statistics is too large, and we never to update it before enforce process start

so, I refactor whole process refs Columnbia paper

1. Add ExploreGroupTask: to explore the chilren of groupExpression
2. Modify DervieStatisticsTask: only dervie self's statistics 
3. Update process: explore all logical plan first, then enforece physical plan, and will update every group's statistics  choose a better one when explore logical plan

Pay attention, the preformance of optimizer may be slower

Explain tpcds sql test:
| TPCDS Plan  |   before(ms) | after(ms) |
| ---- | ---- | ---- |
| Q01     |  640    |  588  |
| Q02     |  177    |  174  |
| Q03     |  46    |  41  |
| Q04     |  296    |  304  |
| Q05     |  134    |  129  |
| Q06     |  46    |  48  |
| Q07     |  36    |  50  |
| Q08     |  55    |  57  |
| Q09     |  76    |  170  |
| Q10     |  72    |  85  |
| Q11     |  75    |  64  |
| Q12     |  43    |  40  |
| Q13     |  76    |  63  |
| Q14_1   |  641    |  317  |
| Q14_2   |  148    |  136  |
| Q15     |  26    |  19  |
| Q16     |  80    |  55  |
| Q17     |  45    |  30  |
| Q18     |  40    |  28  |
| Q19     |  24    |  19  |
| Q20     |  22    |  24  |
| Q21     |  44    |  34  |
| Q22     |  14    |  11  |
| Q23_1   |  148    |  115  |
| Q23_2   |  193    |  174  |
| Q24_1   |  54    |  47  |
| Q24_2   |  42    |  39  |
| Q25     |  30    |  26  |
| Q26     |  15    |  13  |
| Q27     |  20    |  18  |
| Q28     |  36    |  56  |
| Q29     |  25    |  52  |
| Q30     |  39    |  59  |
| Q31     |  88    |  61  |
| Q32     |  59    |  41  |
| Q33     |  105    |  101  |
| Q34     |  22    |  42  |
| Q35     |  52    |  55  |
| Q36     |  32    |  27  |
| Q37     |  19    |  17  |
| Q38     |  22    |  39  |
| Q39_1   |  30    |  52  |
| Q39_2   |  31    |  58  |
| Q40     |  49    |  80  |
| Q41     |  148    |  184  |
| Q42     |  12    |  9  |
| Q43     |  33    |  26  |
| Q44     |  32    |  29  |
| Q45     |  28    |  26  |
| Q46     |  19    |  18  |
| Q47     |  69    |  56  |
| Q48     |  30    |  23  |
| Q49     |  82    |  74  |
| Q50     |  64    |  75  |
| Q51     |  36    |  44  |
| Q52     |  9    |  11  |
| Q53     |  30    |  36  |
| Q54     |  52    |  54  |
| Q55     |  9    |  25  |
| Q56     |  66    |  130  |
| Q57     |  55    |  82  |
| Q58     |  134    |  87  |
| Q59     |  90    |  95  |
| Q60     |  90    |  93  |
| Q61     |  57    |  33  |
| Q62     |  47    |  37  |
| Q63     |  26    |  23  |
| Q64     |  515    |  341  |
| Q65     |  29    |  59  |
| Q66     |  311    |  327  |
| Q67     |  83    |  66  |
| Q68     |  22    |  40  |
| Q69     |  106    |  52  |
| Q70     |  86    |  44  |
| Q71     |  25    |  38  |
| Q72     |  33    |  47  |
| Q73     |  14    |  22  |
| Q74     |  48    |  63  |
| Q75     |  87    |  59  |
| Q76     |  25    |  25  |
| Q77     |  79    |  52  |
| Q78     |  60    |  43  |
| Q79     |  20    |  21  |
| Q80     |  136    |  96  |
| Q81     |  45    |  48  |
| Q82     |  23    |  15  |
| Q83     |  75    |  56  |
| Q84     |  16    |  13  |
| Q85     |  47    |  38  |
| Q86     |  17    |  12  |
| Q87     |  35    |  18  |
| Q88     |  78    |  66  |
| Q89     |  27    |  26  |
| Q90     |  14    |  11  |
| Q91     |  24    |  17  |
| Q92     |  28    |  22  |
| Q93     |  11    |  9  |
| Q94     |  50    |  47  |
| Q95     |  94    |  93  |
| Q96     |  11    |  7  |
| Q97     |  33    |  16  |
| Q98     |  25    |  16  |
| Q99     |  62    |  32  |

http://39.101.206.100:8090/view/benchmark/job/performance_comparison/326/Two_20version_20compare_20report/

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] I have added user document for my new feature or new function

### CODE REVIEW

**Comment by Seaven** [comment]
run all

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 17 / 43 (39.53%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/ExploreGroupTask.java|0|11|00.00%|[15, 16, 17, 21, 22, 25, 26, 27, 29, 30, 34]|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/DeriveStatsTask.java|2|8|25.00%|[33, 34, 35, 46, 47, 48]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Group.java|2|6|33.33%|[77, 81, 82, 191]|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/ApplyRuleTask.java|1|2|50.00%|[92]|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/OptimizeExpressionTask.java|11|15|73.33%|[36, 37, 38, 70]|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/SeriallyTaskScheduler.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 44 / 47 (93.62%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/ApplyRuleTask.java|2|3|66.67%|[49]|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/ExploreGroupTask.java|10|11|90.91%|[34]|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/OptimizeExpressionTask.java|15|16|93.75%|[57]|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/DeriveStatsTask.java|7|7|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Group.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/OptimizeGroupTask.java|4|4|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/task/OptimizeExpressionTask.java [hunk 1]
```diff
@@ -11,49 +11,64 @@
...
-        rules.sort(Comparator.comparingInt(Rule::promise));
 
         for (Rule rule : rules) {
-            pushTask(new ApplyRuleTask(context, groupExpression, rule));
+            pushTask(new ApplyRuleTask(context, groupExpression, rule, isExplore));
+        }
+
+        pushTask(new DeriveStatsTask(context, groupExpression));
+        for (int i = groupExpression.arity() - 1; i >= 0; i--) {
+            pushTask(new ExploreGroupTask(context, groupExpression.getInputs().get(i)));
```
Comment (packy92) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/task/OptimizeExpressionTask.java:71]: Is this a time-consuming place? We will explore the groupExpression and then derive their statistics?

Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/task/OptimizeExpressionTask.java:71]: yeah, explore all child group will increase the time, I will run performance later. we explore all groupExpression for get an accurate statistics

**Comment by kangkaisen** [comment]
Need full performance test and optimizer itself performance test.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/task/OptimizeExpressionTask.java [hunk 2]
```diff
@@ -11,49 +11,64 @@
...
     public String toString() {
-        return "OptimizeExpressionTask for groupExpression " + groupExpression;
+        return "OptimizeExpressionTask" + (this.isExplore ? "[explore]" : "")
+                + " for groupExpression " + groupExpression;
     }
 
     @Override
     public void execute() {
         List<Rule> rules = getValidRules();
-        rules.sort(Comparator.comparingInt(Rule::promise));
```
Comment (liuyehcf) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/task/OptimizeExpressionTask.java:53]: Why is it unnecessary anymore?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/task/OptimizeGroupTask.java [hunk 3]
```diff
@@ -36,12 +35,12 @@ public void execute() {
             return;
         }
 
-        for (GroupExpression expression : group.getLogicalExpressions()) {
-            pushTask(new OptimizeExpressionTask(context, expression));
+        for (int i = group.getLogicalExpressions().size() - 1; i >= 0; i--) {
+            pushTask(new OptimizeExpressionTask(context, group.getLogicalExpressions().get(i)));
```
Comment (liuyehcf) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/task/OptimizeGroupTask.java:39]: Why need to traverse in reverse order?

Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/task/OptimizeGroupTask.java:39]: just for keep enforce logical plan order

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/task/OptimizeExpressionTask.java (hunk 2 above)
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/task/OptimizeExpressionTask.java:53]: move to line 51

**Comment by Seaven** [comment]
> Need full performance test and optimizer itself performance test.

http://39.101.206.100:8090/view/benchmark/job/performance_comparison/326/Two_20version_20compare_20report/

**Comment by Seaven** [comment]
run starrocks_admit_test

**Review by liuyehcf — DISMISSED** [review]

**Review by kangkaisen — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Review by HangyuanLiu — DISMISSED** [review]

**Comment by Seaven** [comment]
run starrocks_admit_test

**Comment by Seaven** [comment]
run starrocks_admit_test

**Comment by Seaven** [comment]
@Mergifyio rebase

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by Seaven** [comment]
run starrocks_fe_unittest

**Comment by Seaven** [comment]
run starrocks_admit_test

**Comment by Seaven** [comment]
run starrocks_fe_unittest

**Comment by Seaven** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 44 / 47 (93.62%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/ApplyRuleTask.java|2|3|66.67%|[49]|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/ExploreGroupTask.java|10|11|90.91%|[34]|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/OptimizeExpressionTask.java|15|16|93.75%|[57]|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/DeriveStatsTask.java|7|7|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Group.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/OptimizeGroupTask.java|4|4|100.00%|[]|

**Pull request merged**